G2-http3-protocol.md

```diff
@@ -0,0 +1,142 @@
+H3_GENERAL_PROTOCOL_ERROR(0x0101)|PROTOCOL_ERROR(1)|INTERNAL
+H3_INTERNAL_ERROR(0x0102)|INTERNAL_ERROR(2)|INTERNAL
+H3_STREAM_CREATION_ERROR(0x0103)|n/a|INTERNAL
+H3_CLOSED_CRITICAL_STREAM(0x0104)|n/a|INTERNAL
+H3_FRAME_UNEXPECTED(0x0105)|FRAME_SIZE_ERROR|INTERNAL
+H3_FRAME_ERROR(0x0106)|FRAME_SIZE_ERROR|INTERNAL
+H3_EXCESSIVE_LOAD(0x0107)|ENHANCE_YOUR_CALM|RESOURCE_EXHAUSTED ...with additional error detail provided by runtime to indicate that the exhausted resource is bandwidth.
+H3_ID_ERROR(0x0108)|n/a|INTERNAL
+H3_SETTINGS_ERROR(0x0109)|SETTINGS_TIMEOUT(4)|INTERNAL
+H3_MISSING_SETTINGS(0x010a)|SETTINGS_TIMEOUT(4)|INTERNAL
+H3_REQUEST_REJECTED(0x010b)|REFUSED_STREAM|UNAVAILABLE - Indicates that no processing occurred and the request can be retried, possibly elsewhere.
+H3_REQUEST_CANCELLED(0x010c)|CANCEL(8)|Mapped to call cancellation when sent by a client.Mapped to CANCELLED when sent by a server. Note that servers should only use this mechanism when they need to cancel a call but the payload byte sequence is incomplete.
+H3_REQUEST_INCOMPLETE(0x010d)|n/a|INTERNAL
+H3_MESSAGE_ERROR(0x010e)|n/a|INTERNAL
+H3_CONNECT_ERROR(0x010f)|CONNECT_ERROR|INTERNAL
+H3_VERSION_FALLBACK(0x0110)|n/a|INTERNAL
+n/a|FLOW_CONTROL_ERROR(3)|INTERNAL
+n/a|STREAM_CLOSED|No mapping as there is no open stream to propagate to. Implementations should log.
+n/a|COMPRESSION_ERROR|INTERNAL
+n/a|INADEQUATE_SECURITY| PERMISSION_DENIED … with additional detail indicating that permission was denied as protocol is not secure enough for call.
+
+#### Connection management
+
+`GOAWAY` and `PING` frames exist in HTTP/3 and serve the same purpose as in HTTP/2.
+One notable difference is the `GOAWAY` frame in HTTP/2 reports the last
+successfully processed stream ID. In HTTP/3 the `GOAWAY` frame ID value must be greater
+that the last successfully processed stream ID.
+
+### Exceeding deadlines
+
+When an RPC has exceeded its deadline in HTTP/2, the server will complete the response
+with a gRPC status code of DEADLINE_EXCEEDED, then send a `RST_STREAM` frame to the
+client to stop it from continuing to send the request stream. The client gets the
+response and reports the DEADLINE_EXCEEDED failure to the app.
+
+The `RST_STREAM` frame doesn't exist in HTTP/3. QUIC supports stopping the reading and sending parts of a QUIC
+stream independently. Two QUIC frames have replaced `RST_STREAM` functionality:
+
+* [RESET_STREAM](https://www.rfc-editor.org/rfc/rfc9000.html#name-reset_stream-frames) - Abruptly terminate the sending part of a stream.
+* [STOP_SENDING](https://www.rfc-editor.org/rfc/rfc9000.html#name-stop_sending-frames) - Requests that a peer cease transmission on a stream.
+
+`RESET_STREAM` is not appropriate because the response with the DEADLINE_EXCEEDED status
+code should reach the client. Because QUIC frames can be received out-of-order,
+sending this QUIC frame after completing the response can cause the client
```

Comment:
Why would this be sent after completing the response? If the server has completed the response, the RPC is over and the deadline no longer matters.

Reply:
I agree. I'm explaining here why RESET_STREAM is the wrong thing to do.

I'm going into detail here because on the surface RESET_STREAM is the first frame someone might try to use as a replacement for HTTP/2's RST_STREAM.

Reply:
I was saying the construction made no sense to me. There is never any point that you'd "send this QUIC frame after completing the response." "Don't do that, but still use RESET_STREAM." "If a deadline occurs, use RESET_STREAM" and yes, that "can cause the client to abort the request" but the client has probably _already_ aborted the request.

Reply:
FYI while you were replying I edit my comment to clarify what I meant by completing a response 😄

Reply:
I don't understand why RESET_STREAM is the wrong thing to use. It seems _at least_ RESET_STREAM should be used, and if the client hasn't half-closed yet then STOP_STREAM as well. That is [even mentioned in the spec](https://www.rfc-editor.org/rfc/rfc9000.html#section-3.5-8):

> An endpoint that wishes to terminate both directions of a bidirectional stream can terminate one direction by sending a RESET_STREAM frame, and it can encourage prompt termination in the opposite direction by sending a STOP_SENDING frame.

If you have to do the `grpc-status: DEADLINE_EXCEEDED` approach, then really we'd also want to allow a truncated message if the status code is not OK. But I don't see what HTTP/2 vs HTTP/3 (or even HTTP/1.1) has to do with that; the reasoning for that approach seems the same in each.

Reply:
I thought with an exceeded deadline situation that it was best practice was for a server to gracefully complete the response by sending trailers with a grpc-status of DEADLINE_EXCEEDED, and then send RST_STREAM. That means if there is a race condition where the response is received before the client timer exceeded the deadline then the correct status (DEADLINE_EXCEEDED) is returned to the user.

If the server sends RESET_STREAM (NO_ERROR) then the client gRPC library will throw an error, and if the client's deadline timer hasn't triggered then they'll probably get an INTERNAL_ERROR status.

Reply:
> I thought with an exceeded deadline situation that it was best practice was for a server to gracefully complete the response by sending trailers with a grpc-status of DEADLINE_EXCEEDED, and then send RST_STREAM

grpc-status closure, followed by RST_STREAM(OK) if client hasn't half-closed, is a thing for normal errors. But we don't use it for when the deadline expires. Instead we use RST_STREAM(CANCEL). C/Go/[Java then have code on the client-side](https://github.com/grpc/grpc-java/blob/4fa612ae3d8c3d38e84184f633fc4fbf02b93a35/core/src/main/java/io/grpc/internal/ClientCallImpl.java#L691-L703) to handle the processing race. And yes, I'd believe that isn't actually documented anywhere. Although truncating the response is mentioned as part of RST_STREAM(CANCEL).

For HTTP/3, the equivalent seems to be RESET_STREAM(H3_REQUEST_CANCELLED) (followed by normal handling if client hasn't half-closed). Not that I know what I'm talking about with H3. I've read like... 5 paragraphs of the spec at this point.

Reply:
I guess I should ask: when the deadline expires and you want to send `grpc-status`, how are you avoiding truncating a message in grpc-dotnet? In Java and Go, the grpc framing is pretty near the application and the network thread just sees raw HTTP body bytes so has no way to synchronize to a message boundary.

Reply:
I don't recall off the top of my head. I'd need to check.

> grpc-status closure, followed by RST_STREAM(OK) if client hasn't half-closed, is a thing for normal errors. But we don't use it for when the deadline expires. Instead we use RST_STREAM(CANCEL). C/Go/Java then have code on the client-side to handle the processing race.

Hmmm, I didn't know that! Looks like grpc-dotnet has been doing deadlines quite differently.

No one has brought up this deadline race as an issue when using the grpc-dotnet client with other implementations. I would have expected to see them if the server is sending back the response you described. Perhaps I added some protection like that. It's been a couple of years since I looked closely at deadline handling logic.

> For HTTP/3, the equivalent seems to be RESET_STREAM(H3_REQUEST_CANCELLED) (followed by normal handling if client hasn't half-closed).

Yes, that matches your description of HTTP/2 deadline behavior.

Reply:
Created https://github.com/grpc/grpc/pull/27216 to improve HTTP/2 documentation. I think what I'm adding there will make the HTTP/3 behaviors to document here more obvious.

Reply:
Fixed.

Also, I made improvements to the grpc-dotnet client so that RST_STREAM(CANCEL) is mapped to DEADLINE_EXCEEDED if received when the client's deadline has been exceeded.